packages/tldraw/src/lib/shapes/text/RichTextArea.tsx

```diff
@@ -0,0 +1,279 @@
+import { EditorView } from '@tiptap/pm/view'
+import { EditorEvents, EditorProvider, JSONContent, type Editor as TTEditor } from '@tiptap/react'
+import {
+	Editor,
+	TLRichText,
+	TLShapeId,
+	preventDefault,
+	stopEventPropagation,
+	useEditor,
+	useUniqueSafeId,
+} from '@tldraw/editor'
+import React, { useCallback, useLayoutEffect, useRef } from 'react'
+
+/** @public */
+export interface TextAreaProps {
+	isEditing: boolean
+	text?: string
+	shapeId: TLShapeId
+	richText?: TLRichText
+	handleFocus(): void
+	handleBlur(): void
+	handleKeyDown(e: KeyboardEvent): void
+	handleChange(changeInfo: { plaintext?: string; richText?: TLRichText }): void
+	handleInputPointerDown(e: React.PointerEvent<HTMLElement>): void
+	handleDoubleClick(e: any): any
+}
+
+/**
+ * N.B. In Development mode you need to ensure you're testing this without StrictMode on.
+ * Otherwise it's not gonna work as expected on iOS.
+ * Specifically, it means that the virtual keyboard won't pop open sometimes
+ * (iOS starts flipping out when you render multiple times when trying to focus something) .
+ */
+
+/**
+ * A rich text area that can be used for editing text with rich text formatting.
+ * This component uses the TipTap editor under the hood.
+ *
+ * @public @react
+ */
+export const RichTextArea = React.forwardRef<HTMLDivElement, TextAreaProps>(function RichTextArea(
+	{
+		shapeId,
+		isEditing,
+		richText,
+		handleFocus,
+		handleChange,
+		handleBlur,
+		handleKeyDown,
+		// TODO: need to handle this still?
+		// handleInputPointerDown,
+		handleDoubleClick,
+	},
+	ref
+) {
+	const editor = useEditor()
+	const tipTapId = useUniqueSafeId('tip-tap-editor')
+	const tipTapConfig = editor.getTextOptions().tipTapConfig
+	const rCreateInfo = useRef({
+		selectAll: false,
+		caretPosition: null as { x: number; y: number } | null,
+		time: Date.now(),
+	})
+
+	const rTextEditor = useRef<TTEditor | null>(null)
+
+	// The order of events is:
+	// - editor begins editing any shape
+	// - we set listeners for select all / place caret events
+	// - if the user is editing this shape, this component is rendered
+	// - editor emits the select all event / place caret event
+	// - the text editor is created
```

Comment:
Rather than rely on this rather complex sequence I was able to use `editor.inputs.originScreenPoint` inside of `handleCreate`. That seems to work? I do think Tiptap's`posAtCoords` assumes full viewport coordinate space, so there might be some complexity when the editor is not rendered edge to edge.

Reply:
The tricky thing is _technically_ there could be another event that fired in the time between `handleCreate` finishes firing, such that `editor.inputs.originScreenPoint` isn't the same point anymore. So, we're trying to make sure the correct point is tied to this event. A bit annoying for sure!